src/main/java/org/springframework/integration/kafka/core/DefaultConnection.java

```diff
@@ -0,0 +1,309 @@
+				long highWatermark = fetchResponse.highWatermark(partition.getTopic(), partition.getId());
+				resultBuilder.add(partition).withResult(new KafkaMessageBatch(partition, kafkaMessages, highWatermark));
+			}
+			else {
+				resultBuilder.add(partition).withError(errorCode);
+			}
+		}
+		return resultBuilder.build();
+	}
+
+	@Override
+	public Result<Long> fetchStoredOffsetsForConsumer(String consumerId, Partition... partitions) throws
+			ConsumerException {
+		FastList<TopicAndPartition> topicsAndPartitions = FastList.newList(Arrays.asList(partitions))
+				.collect(new ConvertToTopicAndPartitionFunction());
+		OffsetFetchRequest offsetFetchRequest = new OffsetFetchRequest(consumerId, topicsAndPartitions,
+				kafka.api.OffsetFetchRequest.CurrentVersion(), createCorrelationId(), simpleConsumer.clientId());
+		OffsetFetchResponse offsetFetchResponse = null;
+		try {
+			offsetFetchResponse = simpleConsumer.fetchOffsets(offsetFetchRequest);
+		}
+		catch (Exception e) {
+			throw new ConsumerException(e);
+		}
+		ResultBuilder<Long> resultBuilder = new ResultBuilder<Long>();
+		for (Partition partition : partitions) {
+			OffsetMetadataAndError offsetMetadataAndError
+					= offsetFetchResponse.offsets().get(new TopicAndPartition(partition.getTopic(), partition.getId()));
+			short errorCode = offsetMetadataAndError.error();
+			if (ErrorMapping.NoError() == errorCode) {
+				resultBuilder.add(partition).withResult(offsetMetadataAndError.offset());
+			}
+			else {
+				resultBuilder.add(partition).withError(errorCode);
+			}
+		}
+		return resultBuilder.build();
+	}
+
+	/**
+	 * @see Connection#fetchInitialOffset(long, Partition...)
+	 */
+	@Override
+	public Result<Long> fetchInitialOffset(long referenceTime, Partition... partitions) throws
+			ConsumerException {
+		Assert.isTrue(partitions.length > 0, "Must provide at least one partition");
+		Map<TopicAndPartition, PartitionOffsetRequestInfo> infoMap = new HashMap<TopicAndPartition, PartitionOffsetRequestInfo>();
+		for (Partition partition : partitions) {
+			infoMap.put(new TopicAndPartition(partition.getTopic(), partition.getId()), new PartitionOffsetRequestInfo(referenceTime, 1));
+		}
+		OffsetRequest offsetRequest = new OffsetRequest(infoMap, kafka.api.OffsetRequest.CurrentVersion(), simpleConsumer.clientId());
+		OffsetResponse offsetResponse = null;
+		try {
+			offsetResponse = simpleConsumer.getOffsetsBefore(offsetRequest);
+		}
+		catch (Exception e) {
+			throw new ConsumerException(e);
+		}
+		ResultBuilder<Long> resultBuilder = new ResultBuilder<Long>();
+		for (Partition partition : partitions) {
+			short errorCode = offsetResponse.errorCode(partition.getTopic(), partition.getId());
+			if (ErrorMapping.NoError() == errorCode) {
+				long[] offsets = offsetResponse.offsets(partition.getTopic(), partition.getId());
+				if (offsets.length == 0) {
+					throw new IllegalStateException("No error has been returned, but no offsets either");
```

Comment:
I don't know Kafka, but this Exception says, that in case of `topic` and `partiotion` exist the `offset` must be present. At least `0`. Correct ?